### PR TITLE
Preserve input order for the returned matching paths

### LINKED ISSUE #14: Preserve list ordering with multiple patterns
Considering: `multimatch(['a/a1', 'b/b', 'a/a2', 'c/c'], ['a/**', 'b/**'])`
I would expect `[ 'a/a1', 'b/b', 'a/a2']`
Instead of `[ 'a/a1', 'a/a2', 'b/b' ]`

If ordering is considered important, it could be offered as an option.

Here's a quick alternative i wrote in coffee. This functionnality is mandatory in my code. I could send you a pull request in js if you are interested, otherwise feel free to discard.

``` coffee
multimatch = (list, patterns, options={}) ->
  list = [list] unless Array.isArray list
  patterns = [patterns] unless Array.isArray patterns
  for el in list
    continue unless patterns.some (pattern) ->
      minimatch el, pattern, options
    el
```

### PR DESCRIPTION
**Pull request by ThanveerulHuq**
Preserve list ordering with multiple patterns #14 issue fixed

Fixes #14

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -66,6 +66,10 @@ test('override negations and re-include explicitly defined patterns', t => {
 	t.deepEqual(multimatch(['foo', 'bar', 'baz'], ['!*a*', '*z']), ['baz']);
 });
 
+test('list be sorted when order enabled', t => {
+	t.deepEqual(multimatch(['foo', 'bar', 'baz'], ['*', '!*{o,r}', 'foo'], {ordered: true}), ['foo', 'baz']);
+});
```
Comment (sindresorhus) [inline test/test.js]: This will need more tests. For example, the case mentioned in the issue: https://github.com/sindresorhus/multimatch/issues/14

File: index.js [hunk 2]
```diff
@@ -22,4 +22,9 @@ module.exports = (list, patterns, options = {}) => {
 
 		return process(result, minimatch.match(list, pattern, options));
 	}, []);
+	if (options.ordered) {
```
Comment (sindresorhus) [inline index.js]: It should not be behind an option.

File: index.js [hunk 3]
```diff
@@ -22,4 +22,9 @@ module.exports = (list, patterns, options = {}) => {
...
+		return list.filter(el => unsorted.indexOf(el) > -1);
```
Comment (sindresorhus) [inline index.js]: Why are you using `filter`? I don't really understand the logic here.

**Comment by sindresorhus** [comment]
The behavior needs to be documented in the readme and index.d.ts.

**Comment by sindresorhus** [comment]
@ThanveerulHuq Friendly bump :)

**Comment by ThanveerulHuq** [comment]
@sindresorhus Added changes based on PR comment

File: index.js [hunk 4]
```diff
@@ -7,24 +7,24 @@ const arrify = require('arrify');
 module.exports = (list, patterns, options = {}) => {
 	list = arrify(list);
 	patterns = arrify(patterns);
+	let result = [];
```
Comment (sindresorhus) [inline index.js:10]: This should be moved to as close as possible to where it's actually used.

File: index.js [hunk 5]
```diff
@@ -7,24 +7,24 @@ const arrify = require('arrify');
...
 
 	if (list.length === 0 || patterns.length === 0) {
 		return [];
 	}
 
-	const unsorted = patterns.reduce((result, pattern) => {
-		let process = arrayUnion;
+	for (const el of list) {
```
Comment (sindresorhus) [inline index.js:16]: `el` needs more descriptive name.

**Comment by sindresorhus** [comment]
> The behavior needs to be documented in the readme and index.d.ts.

⬆️

File: index.d.ts [hunk 6]
```diff
@@ -9,7 +9,7 @@ Extends [`minimatch.match()`](https://github.com/isaacs/minimatch#minimatchmatch
 
 @param paths - Paths to match against.
 @param patterns - Globbing patterns to use. For example: `['*', '!cake']`. See supported [`minimatch` patterns](https://github.com/isaacs/minimatch#usage).
-@returns The matching paths.
+@returns The matching paths in the order input paths.
```
Comment (sindresorhus) [inline index.d.ts:12]: Typo

File: readme.md [hunk 7]
```diff
@@ -32,7 +32,7 @@ Returns an array of matching paths.
 
 Type: `string | string[]`
 
-Paths to match against.
+Paths to match against in the order of input paths.
```
Comment (sindresorhus) [inline readme.md:35]: Not the correct spot

**Comment by sindresorhus** [comment]
The PR could use a better title.

File: readme.md (hunk 7 above)
Comment (ThanveerulHuq) [reply readme.md:35]: No other place fits properly for this info. Can you please suggest a place to add it? or shall I create a new section to add this info?

File: index.d.ts (hunk 6 above)
Comment (ThanveerulHuq) [reply index.d.ts]: fixed

File: readme.md (hunk 7 above)
Comment (sindresorhus) [reply readme.md:35]: I think it's pretty clear if you look at it closer. This is the `path` input variable docs. Your addition is about the return value.

Comment (ThanveerulHuq) [reply readme.md]: fixed

File: readme.md [hunk 8]
```diff
@@ -2,37 +2,34 @@
...
 ```
 $ npm install multimatch
 ```
 
-
 ## Usage
 
 ```js
-const multimatch = require('multimatch');
+const multimatch = require("multimatch");
```
Comment (sindresorhus) [inline readme.md:14]: Don't do unrelated changes.

File: readme.md [hunk 9]
```diff
@@ -2,37 +2,34 @@
...
 
-Returns an array of matching paths.
+Returns an array of matching paths in the order of input paths.
 
 #### paths
 
 Type: `string | string[]`
 
-Paths to match against.
+Paths to match against .
```
Comment (sindresorhus) [inline readme.md:32]: ^

File: readme.md (hunk 8 above)
Comment (ThanveerulHuq) [reply readme.md]: my bad, my formatter introduced this. removed it now
